Tribler/Policies/BoostingSource.py

```diff
@@ -0,0 +1,639 @@
+
+        self.enabled = True
+
+        self.av_uprate = 0
+        self.av_dwnrate = 0
+        self.storage_used = 0
+        self.ready = False
+
+        self.min_connection = boost_settings.min_connection_start
+        self.min_channels = boost_settings.min_channels_start
+
+        self.torrent_mgr = TorrentManagerCM.get_instance(session)
+
+        # local import for handling circular import
+        from Tribler.Policies.BoostingManager import BoostingManager
+        self.boosting_manager = BoostingManager.get_instance()
+        self._logger = logging.getLogger(BoostingManager.__name__)
+
+    def kill_tasks(self):
+        """
+        kill tasks on this source
+
+        *not used right now*
+        """
+        # cancel loading task
+
+        self.ready = False
+        self.torrent_mgr.del_instance()
+
+        self.session.lm.threadpool.cancel_pending_task(str(self.source) + "_load")
+
+        self.session.lm.threadpool.cancel_pending_task(self.source)
+
+    def _load_if_ready(self, source):
+        """
+        load source if and only if the overall system is ready.
+
+        it depends on #connection and #channel
+
+        Useful to not burden the apps in startup
+        """
+
+        nr_channels = self.channelcast_db.getNrChannels()
+        nr_connections = 0
+
+        for community in self.session.lm.dispersy.get_communities():
+            from Tribler.community.search.community import SearchCommunity
+            if isinstance(community, SearchCommunity):
+                nr_connections = community.get_nr_connections()
+
+        # condition example
+        if nr_channels > self.min_channels and nr_connections > self.min_connection:
+            fun = self._load
+        else:
+            fun = self._load_if_ready
+
+        self.session.lm.threadpool.add_task(lambda src=source: fun(src), 15, task_name=str(self.source) + "_load")
+
+    def _load(self, source):
+        pass
+
+    def _update(self):
+        pass
+
+    def get_source(self):
+        """
+        returning 'raw' source. May be overriden
+        """
+        return self.source
+
+
+class ChannelSource(BoostingSource):
+    """
+    Credit mining source from a channel.
+    """
+    def __init__(self, session, dispersy_cid, boost_settings, callback):
+        BoostingSource.__init__(self, session, dispersy_cid, boost_settings, callback)
+
+        self.channel_id = None
+
+        self.channel = None
+        self.community = None
+        self.database_updated = True
+
+        self.session.add_observer(self._on_database_updated, NTFY_TORRENTS, [NTFY_INSERT, NTFY_UPDATE])
+        self.session.lm.threadpool.add_task(lambda cid=dispersy_cid: self._load_if_ready(cid), 0,
+                                            task_name=str(self.source) + "_load")
+
+        self.unavail_torrent = {}
+
+    def kill_tasks(self):
+        BoostingSource.kill_tasks(self)
+
+        if self.community:
+            self.community.cancel_all_pending_tasks()
```

Comment:
Because when running test, the task in `ChannelCommunity` still running. This resulting a dirty reactor. Should I move this to the test code instead of core code?